### PR TITLE
Introduce Factory pattern for building APIHandler

### PR DESCRIPTION
**Pull request by badiib**

### CODE REVIEW

File: cmd/query/app/handler.go [hunk 1]
```diff
@@ -79,18 +82,32 @@ type APIHandler struct {
...
 	}
+
+	for _, option := range options {
+		option(aH)
+	}
+	if aH.httpPrefix == "" {
+		aH.httpPrefix = DefaultHTTPPrefix
+	}
+	if aH.adjuster == nil {
+		aH.adjuster = adjuster.Sequence([]adjuster.Adjuster{
```
Comment (yurishkuro) [inline cmd/query/app/handler.go]: Do we need to create an array? `Sequence()` takes varargs.

File: cmd/query/app/handler.go [hunk 2]
```diff
@@ -47,6 +47,9 @@ const (
 
 	defaultDependencyLookbackDuration = time.Hour * 24
 	defaultTraceQueryLookbackDuration = time.Hour * 24 * 2
+
+	// DefaultHTTPPrefix is the prefix for the API if no prefix is provided
+	DefaultHTTPPrefix = "api"
```
Comment (yurishkuro) [inline cmd/query/app/handler.go]: other default values are private, why does this need to be public?

File: cmd/query/app/handler.go [hunk 3]
```diff
@@ -79,18 +82,32 @@ type APIHandler struct {
...
-// TODO use Builder to allow this package to move to github while accepting custom components like HAProxyMerge()
-func NewAPIHandler(spanReader spanstore.Reader, dependencyReader dependencystore.Reader, logger zap.Logger, httpPrefix string, adjusters []adjuster.Adjuster) *APIHandler {
-	return &APIHandler{
+func NewAPIHandler(spanReader spanstore.Reader, dependencyReader dependencystore.Reader, options ...HandlerOption) *APIHandler {
+	aH := &APIHandler{
 		spanReader:       spanReader,
 		dependencyReader: dependencyReader,
-		logger:           logger,
 		queryParser: queryParser{
 			traceQueryLookbackDuration: defaultTraceQueryLookbackDuration,
```
Comment (yurishkuro) [inline cmd/query/app/handler.go]: there should be an option for this value too

File: cmd/query/app/handler.go (hunk 2 above)
Comment (badiib) [inline cmd/query/app/handler.go]: I was going to use it immediately after in an internal repo.

File: cmd/query/app/handler.go (hunk 3 above)
Comment (badiib) [inline cmd/query/app/handler.go]: +1

File: cmd/query/app/handler.go (hunk 2 above)
Comment (yurishkuro) [inline cmd/query/app/handler.go]: for what purpose? It's already defaulted here, doesn't make sense to re-default. Unless you want to add some other routes to the `/api` prefix, which I don't think is a good idea, extensions should have their own paths.

**Pull request merged**